tasks/market/task_psd.go

```diff
@@ -0,0 +1,362 @@
+package market
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/big"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v9/market"
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/go-state-types/exitcode"
+
+	"github.com/filecoin-project/curio/build"
+	"github.com/filecoin-project/curio/deps/config"
+	"github.com/filecoin-project/curio/harmony/harmonydb"
+	"github.com/filecoin-project/curio/harmony/harmonytask"
+	"github.com/filecoin-project/curio/harmony/resources"
+	"github.com/filecoin-project/curio/lib/multictladdr"
+	"github.com/filecoin-project/curio/lib/promise"
+	"github.com/filecoin-project/curio/tasks/message"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/storage/ctladdr"
+)
+
+const PSDTaskPollInterval = 3 * time.Second
+
+type psdApi interface {
+	ChainHead(context.Context) (*types.TipSet, error)
+	StateMinerInfo(context.Context, address.Address, types.TipSetKey) (api.MinerInfo, error)
+	StateCall(context.Context, *types.Message, types.TipSetKey) (*api.InvocResult, error)
+	ctladdr.NodeApi
+}
+
+type PSDTask struct {
+	db     *harmonydb.DB
+	sender *message.Sender
+	as     *multictladdr.MultiAddressSelector
+	cfg    *config.MK12Config
+	api    psdApi
+
+	TF promise.Promise[harmonytask.AddTaskFunc]
+}
+
+func NewPSDTask(db *harmonydb.DB, sender *message.Sender, as *multictladdr.MultiAddressSelector, cfg *config.MK12Config, api psdApi) *PSDTask {
+	pt := &PSDTask{
+		db:     db,
+		sender: sender,
+		as:     as,
+		cfg:    cfg,
+		api:    api,
+	}
+
+	ctx := context.Background()
+	go pt.pollPSDTasks(ctx)
+	return pt
+}
+
+func (p *PSDTask) pollPSDTasks(ctx context.Context) {
+	for {
+		// Get all deals which do not have a URL and are not after_find
```

Comment:
This poll condition sounds like the last 2. I'm not sure that all these pollers are a good idea. Who is adding to the DB without kicking off the right tasks?

Reply:
Got rid of all these new pollers.